### PR TITLE
Allow an all-zero public input

### PR DESCRIPTION
**Pull request by mrmr1993**
This PR tweaks the handling of public inputs, so that the public input commitment can be allowed to be all-zero.

The handling of zero polynomials and commitments is special-cased in lots of places in the code, so this was a rather tricky change to make. In particular, the 'commitments' test is still broken (via a probabilistic flake), since it depends more closely on some of those special cases.

This is ready for review, but should not be merged until the flaking test is fixed.

### CODE REVIEW

File: poly-commitment/src/evaluation_proof.rs [hunk 1]
```diff
@@ -112,7 +112,7 @@ impl<G: CommitmentCurve> SRS<G> {
             let mut scale = G::ScalarField::one();
 
             // iterating over polynomials in the batch
-            for (p_i, degree_bound, omegas) in plnms.iter().filter(|p| !p.0.is_zero()) {
+            for (p_i, degree_bound, omegas) in plnms.iter() {
```
Comment (mimoo) [inline poly-commitment/src/evaluation_proof.rs:115]: in `&plnms`

File: poly-commitment/src/commitment.rs [hunk 2]
```diff
@@ -435,7 +435,7 @@ pub fn combined_inner_product<F: PrimeField>(
 
         if let Some(m) = shifted {
             // polyscale^i sum_j evalscale^j elm_j^{N - m} f(elm_j)
-            let last_evals = if *m > evals.len() * srs_length {
+            let last_evals = if *m >= evals.len() * srs_length {
```
Comment (mimoo) [inline poly-commitment/src/commitment.rs:438]: I really think we ought to move to this simplified function: https://github.com/o1-labs/proof-systems/pull/448/files and handle chunked polynomials outside of the API

Comment (mimoo) [reply poly-commitment/src/commitment.rs:438]: btw the reason I couldn't move to that code is the empty polynomials (like selectors when you don't use them)

File: poly-commitment/src/commitment.rs [hunk 3]
```diff
@@ -518,15 +518,9 @@ impl<G: CommitmentCurve> SRS<G> {
...
-                    // We should change this so that we still mask in this case
-                    g
-                } else {
-                    let mut g_masked = self.h.mul(b);
-                    g_masked.add_assign_mixed(&g);
-                    g_masked.into_affine()
-                }
+                let mut g_masked = self.h.mul(b);
+                g_masked.add_assign_mixed(&g);
+                g_masked.into_affine()
```
Comment (mimoo) [inline poly-commitment/src/commitment.rs:523]: I'm still not sure how this fixes the issue, as your public polynomial or the selector polynomials are not masked

Comment (jspada) [reply poly-commitment/src/commitment.rs:523]: good question

File: poly-commitment/src/commitment.rs (hunk 2 above)
Comment (mrmr1993) [reply poly-commitment/src/commitment.rs:438]: Perhaps, but I don't think we should block this PR on it.

File: poly-commitment/src/commitment.rs (hunk 3 above)
Comment (mrmr1993) [reply poly-commitment/src/commitment.rs:523]: The tests pass :man_shrugging: This PR makes the handling of zero commitments correct; I'm going to add blinding and propagate it to the pickles side separately.

**Review by mimoo — APPROVED** [review]

File: poly-commitment/src/evaluation_proof.rs [hunk 4]
```diff
@@ -130,9 +130,11 @@ impl<G: CommitmentCurve> SRS<G> {
                     scale *= &polyscale;
                     offset += self.g.len();
                     if let Some(m) = degree_bound {
-                        if offset > *m {
-                            // mixing in the shifted segment since degree is bounded
-                            plnm.add_shifted(scale, self.g.len() - m % self.g.len(), segment);
+                        if offset >= *m {
```
Comment (mimoo) [inline poly-commitment/src/evaluation_proof.rs:133]: perhaps it'd be a good idea to start avoiding "degree" and use `size` or `length`

cf https://twitter.com/cronokirby/status/1568268593946677248

**Pull request merged**